TrafficSimulator/Assets/Car/AutoDrive.cs

```diff
@@ -158,70 +197,147 @@ void Q_UpdateTarget()
                     _status = Status.Driving;
                     
                     // Set the target to the one after the target we missed
-                    _targetIndex = (_repositioningTargetIndex + _repositioningOffset + 1) % _lane.Count;
-                    _target = _lane[_targetIndex];
+                    _target = GetNextLaneNode(_repositioningTarget, _repositioningOffset, false);
                     
                     // Reset the max speed to the original, and set the acceleration to the max again
                     _vehicleController.maxSpeedForward = _originalMaxSpeed;
                     _vehicleController.throttleInput = 1f;
+                    _vehicleController.brakeInput = 0f;
                 }
             }
             // If the vehicle is driving and the target is in front of us and we are close enough
             else if (_status == Status.Driving && dot > 0 && direction.magnitude <= _targetLookaheadDistance)
             {
                 // Set the target to the next point in the lane
-                _targetIndex = (_targetIndex + 1) % _lane.Count;
-                _target = _lane[_targetIndex];
+                _target = GetNextLaneNode(_target, 0, _roadEndBehaviour == RoadEndBehaviour.Loop);
+            }
+
+            // If the vehicle is closer to the target than the brake distance, brake
+            // Also, do not brake at the end node if we are looping
+            if (Vector3.Distance(transform.position, _brakeTarget.Position) <= _brakeDistance && !(_roadEndBehaviour == RoadEndBehaviour.Loop && _brakeTarget == _endNode))
+            {
+                _vehicleController.brakeInput = 1f;
+                _vehicleController.throttleInput = 0f;
+            }
+            // If the vehicle is further away from the target than the brake distance, accelerate
+            else if (Vector3.Distance(transform.position, _brakeTarget.Position) > _brakeDistance)
+            {
+                _vehicleController.brakeInput = 0f;
+                _vehicleController.throttleInput = 1f;
+            }
+        }
+
+        private void Q_UpdateBrakeTarget()
+        {
+            // Set the brake target point to the point closest to the target that is at least _brakeDistance points away
+            while (Vector3.Distance(transform.position, _brakeTarget.Position) < _brakeDistance && _brakeTarget != _endNode)
+            {
+                _brakeTarget = GetNextLaneNode(_brakeTarget, 0, _roadEndBehaviour == RoadEndBehaviour.Stop);
             }
         }
 
-        Vector3 Q_GetTarget()
+        private void Q_UpdateBrakeDistance()
+        {
+            // If we should loop when the road has ended and our braking target is at the end, we do not want any brake offset.
+            // That would cause the vehicle to stop at the end of the road, instead of looping around
+            float brakeOffset = _roadEndBehaviour == RoadEndBehaviour.Loop && _brakeTarget == _endNode ? 0 : _brakeOffset;
```

Comment:
Not sure this is actually how it works. When I implemented the braking the _brakeOffset did not affect any behavior between Loop and Stop mode.